### ISSUE TITLE
random sampling in the latent space of music_attr_vae_reg_gmm

### ISSUE THREAD

**Issue by adrienchaton** [gudgud96/music-fader-nets#2]
Hi !
Thanks for sharing your codes on the music fader nets paper.

I wanted to try it for doing some random sampling, so I followed your notebook and adapted basic functions from my understanding of the model functioning.

def decode_events(model,z,out_length=100):
    out = model.global_decoder(z, steps=out_length)
    pm = magenta_decode_midi(clean_output(out))
    return pm,out

def random_sample(model,CHROMA_DIMS,c_pitch,out_length,c_mode="major",std=1.):
    chroma_dict = {"C": 0, "C#": 1, "D": 2, "D#": 3, "E": 4, "F": 5, "F#": 6, "G": 7, "G#": 8, "A": 9, "A#": 10, "B": 11}
    minor_pos = 12
    c = chroma_dict[c_pitch]
    if c_mode=="minor":
        c += minor_pos
    c = torch.nn.functional.one_hot(torch.tensor([c]),num_classes=CHROMA_DIMS)
    z = torch.cat([torch.normal(0.,std,size=(1, model.latent_dim)), torch.normal(0.,std,size=(1, model.latent_dim)), c], dim=1)
    if torch.cuda.is_available():
        z = z.cuda()
    pm,out = decode_events(model,z,out_length=out_length)
    return pm,out

and the midi outputs are nothing like a melody, or sometimes even getting NaNs

(here is a random pianoroll plot)

[10__major_C#.pdf](https://github.com/gudgud96/music-fader-nets/files/6586118/10__major_C.pdf)

When I run an auto-encode following the process shown in the notebook, I get some more reasonable outputs.

Would you have some insights on how to proceed for random sapling the model please ?

**Comment by gudgud96** [comment]
Hi @adrienchaton,

Sorry for the late reply! Have not been so actively on Github previously, so sorry again!

The short answer is that the prior distribution for z_r and z_n are not standard Gaussians, but GMMs instead. The gist of the paper is to model the distribution of low level features (r and n) as GMMs, with #mixtures = #states of a high-level feature that you wish to correlate (in the paper, it's low & high arousal). 

So you shouldn't be sampling from a standard normal, but you need to first (i) pick arousal state (0 / 1) (ii) find the corresponding mixture's mu and var; (iii) construct a normal distribution with the mu and var. Or else you are likely gonna get an out-of-distribution latent z.

Pseudocode might be sth like:
```
# take low arousal as example, arousal state = 0
mu_r = model.mu_r_lookup(torch.tensor([0]))
var_r = model.logvar_r_lookup(torch.tensor([0])).exp_()
z_r = torch.normal(mu_r, root(var_r)).sample()

# vice-versa for z_n

z = torch.cat([z_r, z_n, c], dim=1)
```

Hope that clarifies!

**Comment by adrienchaton** [comment]
Thanks for the clarification and pseudo-code @gudgud96 , that makes sense :)
